src/branch.c

```diff
@@ -334,9 +334,23 @@ int git_branch_lookup(
 	const char *branch_name,
 	git_branch_t branch_type)
 {
+	int error = -1;
 	assert(ref_out && repo && branch_name);
 
-	return retrieve_branch_reference(ref_out, repo, branch_name, branch_type == GIT_BRANCH_REMOTE);
+	switch (branch_type) {
+	case GIT_BRANCH_LOCAL:
+	case GIT_BRANCH_REMOTE:
+		error = retrieve_branch_reference(ref_out, repo, branch_name, branch_type == GIT_BRANCH_REMOTE);
+		break;
+	case GIT_BRANCH_ALL:
+		error = retrieve_branch_reference(ref_out, repo, branch_name, false);
+		if (error < 0)
+			error = retrieve_branch_reference(ref_out, repo, branch_name, true);
+		break;
+	default:
+		abort();
```

Comment:
`abort` is probably a bit harsh here.  I think `assert(0)` is a reasonable compromise that we've used in the past.